File: completed/96/96.py
import numpy as np
import os

path = os.path.dirname(os.path.abspath(__file__))
datafile = open(path + '\\96_data.txt', 'r')

def possible(y,x,n):
    for i in range(9):
        if int(grid[y][i]) == n: # check row
            return False
    for i in range(9):
        if int(grid[i][x]) == n: # check column
            return False
    cornerX = (x//3)*3
    cornerY = (y//3)*3
    for j in range(3):
        for i in range(3):
            if int(grid[cornerY+j][cornerX+i]) == n: # check square
                return False
    return True

def solve():
    for y in range(9):
        for x in range(9):
            if int(grid[y][x]) == 0:
                for n in range(1, 10):
                    if possible(y,x,n):
                        grid[y][x] = n
                        solve()
                        grid[y][x] = 0
                return 

    corner = int( str(grid[0][0]) + str(grid[0][1]) + str(grid[0][2]) )
    ans_list.append(corner)

lines = datafile.readlines() 
grid = []
row_counter = 0
puzzleID = 1

ans_list = []

for line in lines: 
    if line[0] != "G":
        row = [int(i) for i in list(line.strip())]
        grid.append(row)
        row_counter += 1

    if row_counter == 9:
        solve()
        print("solved puzzle #" + str(puzzleID) + " out of 50")
        puzzleID += 1
        row_counter = 0
        grid = []

a = 0
for item in ans_list:
    a += item

print(a)

File: completed/96/96.cpp
#include <iostream>
#include <fstream>
#include <string>

// std::fstream file("96_data.txt");
// std::string line;
// while (file >> line) {
//     std::cout <<  line << std::endl;   
// }
// file.close();

// int grid1[9][9];

bool possible(int y, int x, int n, int grid[9][9]) {
    // check row
    for (int i = 0; i < 9; i++) {
        if (grid[y][i] == n) { 
            return false;
        }
    }
    // check column
    for (int i = 0; i < 9; i++) {
        if (grid[i][x] == n) { 
            return false;
        }
    }
    // check square
    int cornerX = (x / 3) * 3;
    int cornerY = (y / 3) * 3;
    for (int j = 0; j < 3; j++) {
        for (int i = 0; i < 3; i++) {
            if (grid[cornerY+j][cornerX+i] == n) {
                return false;
            }
        }
    }
    return true;
}

int solve(int grid[9][9]) {
    for (int y = 0; y < 9; y++) {
        for (int x = 0; x < 9; x++) {
            if (grid[y][x] == 0) {
                for (int n = 1; n < 10; n++) {
                    if (possible(y, x, n, grid)) {
                        grid[y][x] = n;
                        solve(grid);
                        grid[y][x] = 0;
                    }
                }
                return 0;
            }
        }
    }
}


int main() {

    int grid[9][9]={{5,3,0,0,7,0,0,0,0},
                    {6,0,0,1,9,5,0,0,0},
                    {0,9,8,0,0,0,0,6,0},
                    {8,0,0,0,6,0,0,0,3},
                    {4,0,0,8,0,3,0,0,1},
                    {7,0,0,0,2,0,0,0,6},
                    {0,6,0,0,0,0,2,8,0},
                    {0,0,0,4,1,9,0,0,5},
                    {0,0,0,0,8,0,0,7,9}};


    // solve(grid);
    int (*ans)[9][9]  = &grid;
    // bool tf = possible(0, 3, 7, grid);

    std::cout << *ans << std::endl;

}